### PR TITLE
perf(rocm): build the AITER decode block table in one kernel, not eleven

### PR DESCRIPTION
**Pull request by demandal25**
## Summary

AITER batch decode rebuilt its dense block table on every `run()` using a chain of ATen ops — ~11 device kernels and ~20 dispatches to feed an attention kernel that is 1-2 launches. The data involved is 128 KB at the largest shape benchmarked, so this was launch-bound, not bandwidth-bound: it showed up as a flat per-call floor that dominated small shapes, where a decode step that should cost 40 µs was paying ~50 µs of dispatch to assemble its own arguments. One fused kernel replaces the chain.

This inverts a finding published in #297. AITER decode was reported there as up to 3.8× slower than the in-tree HIP kernel at small shapes; most of that was this overhead rather than the kernel, and AITER is now **faster** than fa2 at the same shape.

### What changed

- **`include/flashinfer/attention/aiter/batch_decode.cuh`** — new `AiterPaV1BuildBlockTablesKernel` and its launcher, one thread per (seq, slot), grid-strided so the launch config is independent of `max_blocks_per_seq`. The context-length pass is folded into the slot-0 threads rather than costing a second launch. Raw pointers only, per the framework split.
- **`flashinfer/csrc_rocm/batch_decode_aiter.cu`** — `build_block_tables_and_ctxlens` now allocates two `at::empty` outputs and calls the kernel. Both `at::zeros` disappear: the kernel writes every element including the padding, and the two zero tensors that existed only as `at::where` else-operands were never needed — `at::where` takes a scalar.
- **`benchmarks/routines/attention.py`** — MLA's graph-timing gate goes from `!= "fa2"` to `not in ("fa2", "auto")`, matching the three routines #297 updated.

### Architecture / design notes

**Why a flat floor pointed at launch count, not the gather.** The instinct is that a per-call block-table rebuild should scale with `batch × pages_per_seq`, so a *constant* overhead looks like evidence against that explanation. It is not: at batch 64 × s_kv 8192 × page_size 16 the table is 64×512 int32 = 128 KB, and eight passes over it is ~1 MB — sub-microsecond on either architecture. What does not change with shape is that there are ~11 dispatches every time. Launch-bound work produces exactly the flat floor observed.

**The kernel is `static`.** The other `__global__` definitions in `include/` are templates and get weak linkage for free. A non-template one does not, and would collide the moment a second translation unit included this header.

**Iteration space is `max(max_blocks_per_seq, 1)` slots per sequence, not `max_blocks_per_seq`.** With a zero-length grid the slot-0 threads never run, and `context_lens` — now `at::empty` rather than a computed tensor — would be returned uninitialised. The previous code derived it independently of the block table and had no such hazard. Reachable when every sequence in a non-empty batch has zero pages.

**Hoisting the build into `plan()` was considered and deferred.** It would remove the launch entirely rather than collapsing it to one, and it is what fa2 already does — fa2's `plan()` writes its index bookkeeping into the int workspace so `run()` is a single dispatch. But it has to be done with a capacity-sized buffer allocated once and reused: `tests/rocm_tests/test_batch_decode_aiter_hip.py:441` re-calls `plan()` between graph capture and replay, so a fresh allocation per `plan()` would leave the captured graph reading a stale pointer. The fused kernel captures most of the win at a fraction of that risk, and makes the hoist optional rather than urgent.

**Not making `auto` shape-aware.** The obvious alternative was to have the selector prefer fa2 at small shapes. That would have been this repo's first performance-threshold selector — every existing flip is either blanket (`page.py`, `rope.py`, `activation.py`) or a capability gate (`norm.py`'s `ndim == 2`) — and would need per-architecture tuned constants with nothing to keep them honest. It also institutionalises the overhead instead of removing it, and the threshold would be wrong in the other direction the moment this landed.

## Benchmark results

Batch decode, bf16, page_size 16, `--backends fa2 auto`; medians of 5 runs (small) and 3 runs (large). `auto` resolves to AITER for every row here.

**gfx950 / MI350X**, ROCm 7.2.0, torch 2.9.1, amd-aiter 0.1.10:

| shape | before | after | change |
|---|---|---|---|
| 32/8 heads, batch 16, s_kv 1024 | 0.0961 ms ±96.8% | **0.0403 ms ±2.1%** | 2.38× |
| 64/8 heads, batch 64, s_kv 8192 | 0.5507 ms ±0.8% | 0.5183 ms ±0.8% | 1.06× |

**gfx942 / MI300X**, same toolchain:

| shape | before | after | change |
|---|---|---|---|
| 32/8 heads, batch 16, s_kv 1024 | 0.0782 ms | **0.0524 ms** | 1.49× |
| 64/8 heads, batch 64, s_kv 8192 | 0.7293 ms | 0.6957 ms | 1.05× |

fa2 is unchanged on both architectures (1.00×) — it never used this path, which is the control.

**The variance collapse matters as much as the median.** At 96.8% spread the benchmark could not have detected a 40% regression in AITER decode; at 2.1% it can. gfx942 did not show the pathological spread (2.3% before), so that part is CDNA4-specific — but the speedup holds on both. The asymmetry was itself the tell: GPU kernel time does not vary like that, host-side dispatch does.

The gain is largest where the overhead is a bigger fraction of the total, which is small batch and short sequence — i.e. exactly the short-context serving case, and per token per layer.

## Test plan

- [x] `pytest tests/rocm_tests/test_batch_decode_aiter_hip.py -m "not slow"` — **180 passed on gfx950**, **180 passed on gfx942**, zero non-pass marks on either.
- [x] Includes `test_batch_decode_aiter_cuda_graph_replay`, which re-calls `plan()` between capture and replay and is the test that pins the buffer contract any change here has to respect.
- [x] Full ROCm testlist on gfx950 — 46 rows, no errors, unchanged from the merged baseline.
- [x] A/B measured on both architectures against unmodified `amd-integration`, same testlist and same container, only the checkout differing.
- [x] Self-review before push. It caught two defects in this change that the tests did not: the non-template `__global__` in a header, and the uninitialised `context_lens` when `max_blocks_per_seq == 0`. Both fixed before the first push; the tests passed in both states, which is why they needed catching by reading.
- [x] `pre-commit run -a` — clean, no files modified.

## Reviewer notes

The riskiest line is the iteration-space change. `slots_per_seq = max(max_blocks_per_seq, 1)` decouples the grid from the block-table width, so `block_tables` is now indexed as `seq * max_blocks_per_seq + slot` under a `slot < max_blocks_per_seq` guard rather than by the flat loop index. Worth confirming that indexing is right — a wrong stride here produces plausible-looking timings and would only fail numerically.

Worth knowing: the ROCm testlist numbers published in #297 for AITER decode are superseded by this. I'd rather not restate them in the README, since they will move again if the `plan()` hoist lands.

### CODE REVIEW

File: include/flashinfer/attention/aiter/batch_decode.cuh [hunk 1]
```diff
@@ -46,6 +46,68 @@ inline std::size_t AiterPaV1WorkspaceBytes(int num_seqs, int num_heads, int max_
   return per_part * (2 * sizeof(float) + dtype_bytes * static_cast<std::size_t>(head_size));
 }
 
+// Dense block table + context lengths from FlashInfer's flat paged-KV indexing.
+//
+//   block_tables[i][j] = j < npages(i) ? indices[indptr[i] + j] : 0
+//   context_lens[i]    = (npages(i) - 1) * page_size + last_page_len[i]
+//
```
[the remark this answers is not in the document]
Comment (demandal25) [reply include/flashinfer/attention/aiter/batch_decode.cuh]: Both correct, fixed in `5854881d6`.

The underflow arithmetic is inherited unchanged from the ATen implementation, so it is not new — but the previous commit made it *reachable*: extending the iteration space to `max(max_blocks_per_seq, 1)` means the slot-0 threads now run in exactly the all-sequences-empty case that used to early-return. The old behaviour there was uninitialised memory, so neither version was correct. Now clamped to 0, and the documented formula states the arm rather than leaving it to be derived.

File: flashinfer/csrc_rocm/batch_decode_aiter.cu [hunk 2]
```diff
@@ -44,43 +44,28 @@ static at::Tensor get_unit_scale(at::Device device) {
...
+  // empty, not zeros: the kernel writes every slot including the padding.
+  at::Tensor block_tables = at::empty({num_seqs, max_blocks_per_seq}, int32_opts);
+  at::Tensor context_lens = at::empty({num_seqs}, int32_opts);
+
+  const hipStream_t stream = c10::hip::getCurrentHIPStream();
+  const hipError_t status = flashinfer::AiterPaV1BuildBlockTables(
+      paged_kv_indptr.data_ptr<int32_t>(), paged_kv_indices.data_ptr<int32_t>(),
+      paged_kv_last_page_len.data_ptr<int32_t>(), block_tables.data_ptr<int32_t>(),
+      context_lens.data_ptr<int32_t>(), static_cast<int>(num_seqs),
+      static_cast<int>(max_blocks_per_seq), static_cast<int>(page_size), stream);
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/csrc_rocm/batch_decode_aiter.cu]: Valid — the ATen ops absorbed a strided view or a short buffer; `data_ptr<int32_t>()` hands the kernel a base pointer it then indexes as a flat array, and a short `indptr` reads `indptr[seq + 1]` out of bounds. Both silent. Added contiguity checks on all three index tensors plus length checks on `indptr` (num_seqs+1) and `last_page_len` (num_seqs); the dtype checks were already in the caller. Not reachable from the current wrappers, which always pass freshly built contiguous buffers, but cheap and the failure mode is a wrong number rather than a crash. Fixed in `5854881d6`.

**Comment by demandal25** [comment]
Suppressed-comment disposition for the Copilot review on `5854881d6` (1 comment): the `batch_decode.cuh:86` `context_lens` underflow is the same finding as the inline comment on line 53 and is fixed in the same commit — clamped to 0 when `npages == 0`, with the documented formula updated to state the arm.

**Comment by demandal25** [comment]
Latest round addressed in `978a7e1a4`: the stale `context_lens` formula in the `.cu` file comment (the header copy was updated in `5854881d6`, this one was missed), and a device check on the three index tensors — the third precondition the ATen chain enforced for free, alongside the contiguity and length checks already added.

On the suppressed comment asking for a test of the all-zero-page case: I wrote one and dropped it rather than ship it. It cannot do its job. An all-zero-page batch never reaches the kernel — `max_context_len` is 0, so AITER's codegen fails at `plan()` with `npar_loops=0`. A mixed batch does reach it, but passes identically with and without the clamp. And when it did fail it failed order-dependently — green in isolation, red in the full file — on zero-page rows returning **non-finite** values from AITER *with* the clamp applied. That is a pre-existing AITER limitation rather than anything this branch introduced, and it deserves its own investigation, not a flaky assertion here. The clamp stays as defence in depth, explicitly documented as untested.

**Comment by demandal25** [comment]
Swept the suppressed blocks across all reviews on this PR (they are not cumulative, so each review body needs reading separately). Disposition of everything raised there:

| Raised in | Finding | Status |
|---|---|---|
| 5022059797 | `context_lens` underflows at `npages == 0` | Fixed, `5854881d6` |
| 5022945032 | No test for the all-empty-pages case | Won't fix — explained below |
| 5023020864 | Stale `context_lens` formula in the `.cu` comment | Fixed, `978a7e1a4` |
| 5023020864 / 5023083620 | No device check on the index tensors | Fixed, `978a7e1a4` |
| 5023020864 | Kernel needs `indices_numel` threaded through for a bounds check | **Fixed, `b624cd053`** — this was the one genuinely open item |

The bounds check is the substantive one and I had missed it. The ATen `index_select` validated its gather indices; reading `indices[begin + slot]` directly does not, so an `indptr` whose last element exceeds `paged_kv_indices.numel()` walks off the buffer instead of raising. `indptr` is on the device so this cannot be checked host-side without a sync — the extent is threaded into the kernel and out-of-range slots fall back to 0, which is the padding they would have received anyway. That completes the set of preconditions the ATen chain gave for free: contiguity and length (`5854881d6`), device (`978a7e1a4`), index extent (`b624cd053`).

On the missing test: I wrote one and dropped it rather than ship it, for the reasons in the previous comment — an all-zero-page batch cannot reach the kernel (`plan()` fails first in AITER's codegen with `npar_loops=0`), and a mixed batch passes identically with and without the clamp. The clamp and the `slots_per_seq` guard stay as defence in depth, documented as untested rather than covered by an assertion that proves nothing.

180 passed on gfx950 and 180 on gfx942 at `b624cd053`.

File: include/flashinfer/attention/aiter/batch_decode.cuh [hunk 3]
```diff
@@ -46,6 +46,73 @@ inline std::size_t AiterPaV1WorkspaceBytes(int num_seqs, int num_heads, int max_
...
+  for (int64_t idx = blockIdx.x * static_cast<int64_t>(blockDim.x) + threadIdx.x; idx < total;
+       idx += static_cast<int64_t>(gridDim.x) * blockDim.x) {
+    const int seq = static_cast<int>(idx / slots_per_seq);
+    const int slot = static_cast<int>(idx % slots_per_seq);
+    const int32_t begin = indptr[seq];
+    const int32_t npages = indptr[seq + 1] - begin;
+
+    if (slot < max_blocks_per_seq) {
+      // The ATen index_select this replaced bounds-checked the gather; reading
+      // indices[] directly means a malformed indptr walks off the end instead.
```
[the remark this answers is not in the document]
Comment (demandal25) [reply include/flashinfer/attention/aiter/batch_decode.cuh]: Fair — the comment described the hazard rather than the behaviour, and it read as if the OOB were still live. Rewritten to say what the code does and to name the tradeoff explicitly: `indptr` is device memory, so validating it host-side would cost a sync on every `run()`, and a defined value is preferred to an OOB read at the price of masking a malformed `indptr` rather than raising on it. Fixed in `6837ef7a9`.

File: flashinfer/csrc_rocm/batch_decode_aiter.cu [hunk 4]
```diff
@@ -42,45 +41,42 @@ static at::Tensor get_unit_scale(at::Device device) {
...
+  // cannot absorb a strided view or a short buffer -- either would read OOB.
+  TORCH_CHECK(paged_kv_indptr.is_contiguous() && paged_kv_indices.is_contiguous() &&
+                  paged_kv_last_page_len.is_contiguous(),
+              "AITER PA v1 requires contiguous paged-KV index tensors");
+  TORCH_CHECK(paged_kv_indptr.numel() == num_seqs + 1, "paged_kv_indptr must have num_seqs+1 (",
+              num_seqs + 1, ") elements; got ", paged_kv_indptr.numel());
+  TORCH_CHECK(paged_kv_last_page_len.numel() == num_seqs,
+              "paged_kv_last_page_len must have num_seqs (", num_seqs, ") elements; got ",
+              paged_kv_last_page_len.numel());
+
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/csrc_rocm/batch_decode_aiter.cu]: Added the 1-D checks in `edfbc9ce7` — the contract is worth enforcing and the ATen path did reject 2-D outright, so keeping that rejection is right.

Correcting the rationale though: this cannot produce incorrect results. A contiguous tensor with the same `numel` has the same flat layout, so a `[1, total_pages]` indices tensor reads byte-identically to `[total_pages]` through `data_ptr<int32_t>()`. Checked (12,), (1,12), (12,1) and (3,4) — the flat read matches in every case. So the change is contract hygiene, not a silent-miscomputation fix.
